<html>

<head>
<title>Entering a gene list</title>
<style type="text/css">
	body {
		font-family: sans-serif;
	}
	li {
	padding-bottom: 1em
	}
 </style>
</head>
<body>
<h1>Entering a list of genes</h1>
<h2>Query genes</h2>
<p>
A list of human or mouse genes symbols can be pasted in to the query gene field.
</p>
<br>
<h2>Background genes</h2>
<p>
By default all the genes in the genome (from a gtf file) are used as the background set.
Depending on the experiment that the gene set was derived from, this may or may not be the most appropriate set of background genes. 
Custom background genes can be entered by changing the dropdown box from 'Use all genes in genome' 
to 'Enter custom background genes'. The background genes can then be pasted in. Again, these must be gene symbols.  
</p>
<br>
<h2>Statistical test</h2>
To test for enrichment of the set of query genes in the functional categories, a Fisher's Exact test is performed for each category.
A Benjamini-Hochberg multiple testing correction is applied. Further reference to p-values refers to the B-H corrected p-values. 

<h2>Setting options</h2>
<h3>p-value</h3>
The p-value threshold can be adjusted before the data is processed - this may be preferable if the data produces a large number of 
significant categories. It can be adjusted further once the data has been loaded and processed. 
<h3>Number of genes in functional category</h3>
This refers to the total number of genes contained within a functional category. 
Some of the categories are very broad and refer to generic biological processes; these will often contain many hundreds or thousands of genes.
Other functional categories can be very specific and contain as few members as 1 gene. At the extremes these categories aren't very useful so it is preferable to filter these out.

</body>
</html>

